pydantic/generics.py

```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
         """
         This method can be overridden to achieve a custom naming scheme for GenericModels
         """
-        param_names = [param.__name__ if hasattr(param, '__name__') else str(param) for param in params]
+        try:
+            GenericAlias = typing.GenericAlias  # type: ignore
+        except AttributeError:
+            GenericAlias = ()  # for all python < 3.9
+        param_names = [
```

Comment:
there's a lot going on in this comprehension. Might be more readable to pull it out into a boring old.

```py
param_names = []
for param in params:
    if ...:
        param_names.append(...)
```

Reply:
also please add comments explaining what's going on here.

Reply:
See below, got rid of the weird stuff with `pydantic.typing.display_as_type`